util/slack_wrapper.py

```diff
@@ -1,202 +1,223 @@
 import json
-import time
+from typing import Any, List, Dict, Union
+
+from slack_sdk import WebClient
+from slack_sdk.socket_mode.request import SocketModeRequest
+from slack_sdk.socket_mode.response import SocketModeResponse
+from slack_sdk.web import SlackResponse
+from slack_sdk.socket_mode import SocketModeClient
 
-from slackclient import SlackClient
 from util.util import load_json
 
 
 class SlackWrapper:
     """
     Slack API wrapper
+
+    Does not validate responses!
     """
 
-    def __init__(self, api_key):
+    def __init__(self, slack_token, socket_mode_token, handler):
         """
         SlackWrapper constructor.
         Connect to the real-time messaging API and
         load the bot's login data.
         """
-        self.api_key = api_key
-        self.client = SlackClient(self.api_key)
-        self.connected = self.client.rtm_connect(auto_reconnect=True)
-        self.server = None
-        self.username = None
-        self.user_id = None
+        self.client = WebClient(slack_token)
+
+        identity = self.client.auth_test()
+        identity.validate()
+
+        self.user_id = identity['user_id']
+        self.username = identity['user']
+
+        self.on_message_handlers = [handler]
+        self.seen_messages = []
+
+        self.socket = SocketModeClient(socket_mode_token)
+        self.socket.connect()
+        self.socket.socket_mode_request_listeners.append(self.handle)
```

Comment:
the handler is being registered twice and causing retransmission. getting rid of line 34 or line 39 should fix.

Reply:
Probably a bug!

Reply:
I haven't actually been able to repro it since I saw it happen, so something strange may have been happening related to my local config.